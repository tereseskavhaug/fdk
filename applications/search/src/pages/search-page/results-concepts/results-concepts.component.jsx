import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Button, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import cx from 'classnames';
import _get from 'lodash/get';
import _capitalize from 'lodash/capitalize';
import _ from 'lodash';
import { Link } from 'react-router-dom';

import {
  PATHNAME_CONCEPTS,
  PATHNAME_CONCEPTS_COMPARE
} from '../../../constants/constants';
import localization from '../../../lib/localization';
import { ConceptsHitItem } from './concepts-hit-item/concepts-hit-item.component';
import { CompareTerms } from './compare-terms/compare-terms.component';
import { SearchPublishersTree } from '../search-publishers-tree/search-publishers-tree.component';
import { getTranslateText } from '../../../lib/translateText';

export class ResultsConcepts extends React.Component {
  componentWillMount() {
    const urlHasSortfieldModified =
      window.location.href.indexOf('sortfield=modified') !== -1;
    const conceptSortValueIsModified =
      this.props.conceptSortValue === 'modified';

    if (conceptSortValueIsModified || urlHasSortfieldModified) {
      this.props.setConceptSort('modified');
      this.props.onSortByLastModified();
    } else {
      this.props.onSortByScore();
      this.props.setConceptSort(undefined);
    }
  }
  _renderCompareTerms() {
    const { conceptsCompare } = this.props;
    const conceptIdsArray = [];
    const children = items =>
      Object.keys(items).map(el => {
        const item = items[el];
        conceptIdsArray.push(item.id);
        const { publisher } = item;
        const publisherPrefLabel =
          getTranslateText(_get(publisher, ['prefLabel'])) ||
          _capitalize(_get(publisher, 'name', ''));

        return (
          <CompareTerms
            key={item.uri}
            uri={item.id}
            prefLabel={item.prefLabel}
            creator={publisherPrefLabel}
            onDeleteTerm={this.props.removeConcept}
          />
        );
      });

    if (conceptsCompare && Object.keys(conceptsCompare).length > 0) {
      return (
        <div className="mt-5">
          <h3 className="mb-3">{localization.terms.compareTerms}</h3>
          {children(conceptsCompare)}
          <div className="d-flex justify-content-center">
            <Link
              to={`${PATHNAME_CONCEPTS}${PATHNAME_CONCEPTS_COMPARE}?compare=${conceptIdsArray}`}
            >
              {localization.compare.openCompare}
            </Link>
          </div>
        </div>
      );
    }
    return null;
  }

  _renderTerms() {
    const { conceptItems, conceptsCompare } = this.props;
    if (conceptItems && Array.isArray(conceptItems)) {
      return conceptItems.map(item => (
        <ConceptsHitItem
          key={item.id}
          result={item}
          concepts={conceptsCompare}
          onAddConcept={this.props.addConcept}
          onDeleteConcept={this.props.removeConcept}
        />
      ));
    }
    return null;
  }

  _renderFilterModal() {
    const {
      showFilterModal,
      closeFilterModal,
      onFilterPublisherHierarchy,
      searchQuery,
      publisherArray,
      publishers
    } = this.props;
    return (
      <Modal isOpen={showFilterModal} toggle={closeFilterModal}>
        <ModalHeader toggle={closeFilterModal}>Filter</ModalHeader>
        <ModalBody>
          <div className="search-filters">
            <SearchPublishersTree
              title={localization.facet.organisation}
              filter={publisherArray}
              onFilterPublisherHierarchy={onFilterPublisherHierarchy}
              activeFilter={searchQuery.orgPath}
              publishers={publishers}
            />
          </div>
        </ModalBody>
        <ModalFooter>
          <Button
            className="fdk-button-default fdk-button"
            onClick={closeFilterModal}
          >
            Close
          </Button>
        </ModalFooter>
      </Modal>
    );
  }

  render() {
    const {
      conceptAggregations,
      conceptTotal,
      onClearFilters,
      onPageChange,
      onFilterPublisherHierarchy,
      searchQuery,
      showClearFilterButton,
      hitsPerPage,
      publisherArray,
      publishers,
      onSortByScore,
      onSortByLastModified,
      conceptSortValue,
      setConceptSort
    } = this.props;
    const page = _.get(searchQuery, 'from')
      ? searchQuery.from / hitsPerPage
      : 0;
    const pageCount = Math.ceil((conceptTotal || 1) / hitsPerPage);
    const clearButtonClass = cx(
      'btn',
      'btn-primary',
      'fdk-button',
      'fade-in-500',
      {
        'd-none': !showClearFilterButton
      }
    );
    const sortByScoreClass = cx('fdk-button', 'fdk-button-black-toggle', {
      selected: conceptSortValue === undefined
    });
    const sortByLastModifiedClass = cx(
      'fdk-button',
      'fdk-button-black-toggle',
      {
        selected: conceptSortValue === 'modified'
      }
    );

    const onSortByScoreClick = () => {
      setConceptSort(undefined);
      onSortByScore();
    };
    const onSortByModifiedClick = () => {
      setConceptSort('modified');
      onSortByLastModified();
    };

    return (
      <main id="content">
        <section className="row mb-3 fdk-button-row">
          <div className="col-6 col-lg-4">
            <button
              className={clearButtonClass}
              onClick={onClearFilters}
              type="button"
            >
              {localization.query.clear}
            </button>
          </div>
          <div className="col-6 col-lg-4 offset-lg-4">
            <div className="d-flex justify-content-end">
              <Button
                className={sortByScoreClass}
                onClick={onSortByScoreClick}
                color="primary"
              >
                {localization.sort.relevance}
              </Button>
              <Button
                className={sortByLastModifiedClass}
                onClick={onSortByModifiedClick}
                color="primary"
              >
                {localization.sort.modified}
              </Button>
            </div>
          </div>
        </section>

        <section className="row">
          <aside className="search-filters col-lg-4">
            <div className="d-none d-lg-block">
              <span className="uu-invisible" aria-hidden="false">
                Filtrering tilgang
              </span>
              {conceptAggregations && (
                <div>
                  {this._renderFilterModal()}
                  <SearchPublishersTree
                    title={localization.facet.organisation}
                    filter={publisherArray}
                    onFilterPublisherHierarchy={onFilterPublisherHierarchy}
                    activeFilter={_.get(searchQuery, 'orgPath')}
                    publishers={publishers}
                  />
                </div>
              )}
            </div>
            {this._renderCompareTerms()}
          </aside>

          <section className="col-lg-8">{this._renderTerms()}</section>

          {conceptTotal > 50 && (
            <section className="col-lg-8 offset-lg-4 d-flex justify-content-center">
              <span className="uu-invisible" aria-hidden="false">
                Sidepaginering.
              </span>
              <ReactPaginate
                pageCount={pageCount}
                pageRangeDisplayed={2}
                marginPagesDisplayed={1}
                previousLabel={localization.page.prev}
                nextLabel={localization.page.next}
                breakLabel={<span>...</span>}
                breakClassName="break-me"
                containerClassName="pagination"
                onPageChange={onPageChange}
                subContainerClassName="pages pagination"
                activeClassName="active"
                initialPage={page}
                disableInitialCallback
              />
            </section>
          )}
        </section>
      </main>
    );
  }
}

ResultsConcepts.defaultProps = {
  conceptItems: null,
  onClearFilters: null,
  onPageChange: null,
  onFilterPublisherHierarchy: null,
  searchQuery: null,
  hitsPerPage: null,
  showFilterModal: null,
  closeFilterModal: null,
  showClearFilterButton: null,
  publisherArray: null,
  publishers: null,
  conceptsCompare: null,
  addConcept: _.noop,
  setConceptSort: null
};

ResultsConcepts.propTypes = {
  conceptItems: PropTypes.array,
  onClearFilters: PropTypes.func,
  onPageChange: PropTypes.func,
  onFilterPublisherHierarchy: PropTypes.func,
  searchQuery: PropTypes.object,
  hitsPerPage: PropTypes.number,
  showFilterModal: PropTypes.bool,
  closeFilterModal: PropTypes.func,
  showClearFilterButton: PropTypes.bool,
  publisherArray: PropTypes.array,
  publishers: PropTypes.object,
  conceptsCompare: PropTypes.object,
  addConcept: PropTypes.func,
  setConceptSort: PropTypes.func,
  onSortByLastModified: PropTypes.func.isRequired,
  onSortByScore: PropTypes.func.isRequired
};
