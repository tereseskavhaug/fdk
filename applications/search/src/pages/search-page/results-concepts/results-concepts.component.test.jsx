import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import { ResultsConcepts } from './results-concepts.component';
import concepts from '../../../../test/fixtures/concepts';
import { HITS_PER_PAGE } from '../../../constants/constants';
import { normalizeAggregations } from '../../../lib/normalizeAggregations';

let onClearFilters;
let onPageChange;
let onFilterPublisherHierarchy;
let closeFilterModal;
let setConceptSort;
let onSortByLastModified;
let onSortByScore;
let defaultProps;
let wrapper;

beforeEach(() => {
  onClearFilters = jest.fn();
  onPageChange = jest.fn();
  onFilterPublisherHierarchy = jest.fn();
  closeFilterModal = jest.fn();
  setConceptSort = jest.fn();
  onSortByLastModified = jest.fn();
  onSortByScore = jest.fn();

  defaultProps = {
    termItems: concepts,
    onClearFilters,
    onPageChange,
    onFilterPublisherHierarchy,
    searchQuery: {},
    hitsPerPage: HITS_PER_PAGE,
    showFilterModal: false,
    closeFilterModal,
    showClearFilterButton: false,
    publisherArray: null,
    publishers: null,
    setConceptSort,
    onSortByLastModified,
    onSortByScore
  };
  wrapper = shallow(<ResultsConcepts {...defaultProps} />);
  wrapper.setState({
    terms: [concepts.hits.hits[0]._source]
  });
});

test('should render ResultsConcepts correctly with minimum of props', () => {
  const minWrapper = shallow(
    <ResultsConcepts
      setConceptSort={setConceptSort}
      onSortByLastModified={onSortByLastModified}
      onSortByScore={onSortByScore}
    />
  );
  expect(minWrapper).toMatchSnapshot();
});

test('should render ResultsConcepts correctly with props', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render ResultsConcepts correctly with hits', () => {
  wrapper.setProps({
    conceptItems: _.get(concepts, ['_embedded', 'concepts']),
    conceptAggregations: _.get(normalizeAggregations(concepts), 'aggregations'),
    conceptTotal: _.get(concepts, ['page', 'totalElements'])
  });
  expect(wrapper).toMatchSnapshot();
});
