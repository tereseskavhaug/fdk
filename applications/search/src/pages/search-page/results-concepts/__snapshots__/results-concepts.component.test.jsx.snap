// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ResultsConcepts correctly with hits 1`] = `
<main
  id="content"
>
  <section
    className="row mb-3 fdk-button-row"
  >
    <div
      className="col-6 col-lg-4"
    >
      <button
        className="btn btn-primary fdk-button fade-in-500 d-none"
        onClick={[MockFunction]}
        type="button"
      >
        Clear
      </button>
    </div>
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </section>
  <section
    className="row"
  >
    <aside
      className="search-filters col-lg-4"
    >
      <div
        className="d-none d-lg-block"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Filtrering tilgang
        </span>
        <div>
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransition={
              Object {
                "mountOnEnter": true,
                "timeout": 150,
              }
            }
            centered={false}
            fade={true}
            isOpen={false}
            keyboard={true}
            modalTransition={
              Object {
                "timeout": 300,
              }
            }
            onClosed={[Function]}
            onOpened={[Function]}
            role="dialog"
            toggle={[MockFunction]}
            zIndex={1050}
          >
            <ModalHeader
              closeAriaLabel="Close"
              tag="h5"
              toggle={[MockFunction]}
              wrapTag="div"
            >
              Filter
            </ModalHeader>
            <ModalBody
              tag="div"
            >
              <div
                className="search-filters"
              >
                <SearchPublishersTree
                  activeFilter={null}
                  filter={null}
                  onFilterPublisherHierarchy={[MockFunction]}
                  onSearch={null}
                  orgPath={null}
                  publishers={null}
                  title="Organization"
                />
              </div>
            </ModalBody>
            <ModalFooter
              tag="div"
            >
              <Button
                className="fdk-button-default fdk-button"
                color="secondary"
                onClick={[MockFunction]}
                tag="button"
              >
                Close
              </Button>
            </ModalFooter>
          </Modal>
          <SearchPublishersTree
            activeFilter={null}
            filter={null}
            onFilterPublisherHierarchy={[MockFunction]}
            onSearch={null}
            orgPath={null}
            publishers={null}
            title="Organization"
          />
        </div>
      </div>
    </aside>
    <section
      className="col-lg-8"
    />
  </section>
</main>
`;

exports[`should render ResultsConcepts correctly with minimum of props 1`] = `
<main
  id="content"
>
  <section
    className="row mb-3 fdk-button-row"
  >
    <div
      className="col-6 col-lg-4"
    >
      <button
        className="btn btn-primary fdk-button fade-in-500 d-none"
        onClick={null}
        type="button"
      >
        Clear
      </button>
    </div>
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </section>
  <section
    className="row"
  >
    <aside
      className="search-filters col-lg-4"
    >
      <div
        className="d-none d-lg-block"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Filtrering tilgang
        </span>
      </div>
    </aside>
    <section
      className="col-lg-8"
    />
  </section>
</main>
`;

exports[`should render ResultsConcepts correctly with props 1`] = `
<main
  id="content"
>
  <section
    className="row mb-3 fdk-button-row"
  >
    <div
      className="col-6 col-lg-4"
    >
      <button
        className="btn btn-primary fdk-button fade-in-500 d-none"
        onClick={[MockFunction]}
        type="button"
      >
        Clear
      </button>
    </div>
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </section>
  <section
    className="row"
  >
    <aside
      className="search-filters col-lg-4"
    >
      <div
        className="d-none d-lg-block"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Filtrering tilgang
        </span>
      </div>
    </aside>
    <section
      className="col-lg-8"
    />
  </section>
</main>
`;
